crates/chain/src/local_chain.rs

```diff
@@ -19,215 +130,295 @@ impl ChainOracle for LocalChain {
         }
     }
 
-    /// Get a reference to a map of block height to hash.
-    pub fn blocks(&self) -> &BTreeMap<u32, BlockHash> {
-        &self.blocks
+    /// Get the highest checkpoint.
+    pub fn tip(&self) -> Option<CheckPoint> {
+        self.checkpoints.values().last().cloned()
     }
 
-    /// Get the chain tip.
-    pub fn tip(&self) -> Option<BlockId> {
-        self.blocks
-            .iter()
-            .last()
-            .map(|(&height, &hash)| BlockId { height, hash })
+    /// Returns whether the [`LocalChain`] is empty (has no checkpoints).
+    pub fn is_empty(&self) -> bool {
+        self.checkpoints.is_empty()
     }
 
-    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
-    /// are to be re-filled.
-    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet, UpdateNotConnectedError> {
-        let update = update.as_ref();
-        let update_tip = match update.keys().last().cloned() {
-            Some(tip) => tip,
-            None => return Ok(ChangeSet::default()),
-        };
-
-        // this is the latest height where both the update and local chain has the same block hash
-        let agreement_height = update
-            .iter()
-            .rev()
-            .find(|&(u_height, u_hash)| self.blocks.get(u_height) == Some(u_hash))
-            .map(|(&height, _)| height);
+    /// Updates [`Self`] with the given `new_tip`.
+    ///
+    /// The method returns [`ChangeSet`] on success. This represents the applied changes to
+    /// [`Self`].
+    ///
+    /// To update, the `new_tip` must *connect* with `self`. If `self` and `new_tip` has a mutual
+    /// checkpoint (same height and hash), it can connect if:
+    /// * The mutual checkpoint is the tip of `self`.
+    /// * An ancestor of `new_tip` has a height which is of the checkpoint one higher than the
+    ///         mutual checkpoint from `self`.
+    ///
+    /// Additionally:
+    /// * If `self` is empty, `new_tip` will always connect.
+    /// * If `self` only has one checkpoint, `new_tip` must have an ancestor checkpoint with the
+    ///     same height as it.
+    ///
+    /// To invalidate from a given checkpoint, `new_tip` must contain an ancestor checkpoint with
+    /// the same height but different hash.
+    ///
+    /// # Errors
+    ///
+    /// An error will occur if the update does not correctly connect with `self`.
+    ///
+    /// Refer to [module-level documentation] for more.
+    ///
+    /// [module-level documentation]: crate::local_chain
+    pub fn update(&mut self, new_tip: CheckPoint) -> Result<ChangeSet, CannotConnectError> {
+        let mut updated_cps = BTreeMap::<u32, CheckPoint>::new();
+        let mut agreement_height = Option::<u32>::None;
+        let mut agreement_ptr_matches = false;
+
+        for cp in new_tip.iter() {
+            let block = cp.block_id();
+
+            match self.checkpoints.get(&block.height) {
```

Comment:
I think this algorithm does not require us to use `BTreeMap`s at all. We are just trying to figure out whether and where two linked lists connect to each other with our given chain rules. This can be done by traversing the linked lists backwards in tandem. The `BTreeMap` should just be there to index the linked list items by height once we've figured out our new chain.

I don't mind implementing it like this if you don't see my vision here.

Reply:
@LLFourn today I tried implementing it the way as you suggested. However, the algorithm to traverse the linked lists in tandem is difficult (I will not claim that it is impossible).

One of the reasons that makes this a difficult problem is that our update logic requires a specific rule which states:

> The update chain must have a checkpoint that overlaps the original chain's checkpoint which is one higher than the highest point of agreement.

However, depending on where we have our points of agreements (or lack of), we will need to have to include/exclude specific ranges of blocks from the original and update chains, while ensuring we don't unnecessarily remove links for the final resultant chain.

Essentially, it is much easier to not care about links until the final step (hence the `fix_links` method).

Would you be okay to keep this as is and we fix `LocalChain` for good in #1005?

Reply:
@LLFourn has been persistent to have this change (and for good reason) since it has simplified our implementation (see the `local_chain::merge_chains` method). These changes are included in this PR.